File: 31-next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        // Optimal Approach: 

        // int n = nums.size();
        // int ind = -1;

        // //Finding the Breakpoint for the longest-Prefix Match:
        // for(int i = n-2; i>=0; i--){
        //     if(nums[i]<nums[i+1]){
        //         ind = i;
        //         break;
        //     }
        // }

        // if(ind==-1){
        //     reverse(nums.begin(), nums.end());
        //     // for(auto& i: nums) cout<<i<<" ";
        //     return;
        // }

        // for(int i = n-1; i>ind; i--){
        //     if(nums[i]>nums[ind]){
        //         swap(nums[i], nums[ind]);
        //         break;
        //     }
        // }

        // sort(nums.begin()+ind+1, nums.end());


        //Using STL:
        next_permutation(nums.begin(), nums.end());
        return;
        
    }
};

File: 31-next-permutation/next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        n = len(nums)
        critical = -1
        for i in range(n-1, 0, -1):
            if nums[i]>nums[i-1]:
                critical = i
                break

        if critical==-1:
           nums.sort()
           return
        
        # Swap the next element with the smallest greater element than this one:

        for i in range(n-1, critical-1, -1):
            if nums[i]>nums[critical-1]:
                nums[i], nums[critical-1] = nums[critical-1], nums[i]
                break

        #Sorting the remaining elements after the critical:
        nums[critical:] = sorted(nums[critical:])